src/goto-programs/abstract-interpretation/gcse.cpp

```diff
@@ -0,0 +1,468 @@
+  const cse_domaint &b,
+  goto_programt::const_targett from,
+  goto_programt::const_targett to)
+{
+  /* This analysis is supposed to be used for CFGs.
+   * Since we do not have a CFG GOTO abstract interpreter we
+   * simulate one by just passing through common instructions
+   * and only doing intersections at target destinations */
+
+  if(!(to->is_target() || from->is_function_call()) || is_bottom())
+  {
+    bool changed = available_expressions != b.available_expressions;
+    available_expressions = b.available_expressions;
+    return changed;
+  }
+
+  size_t size_before_intersection = available_expressions.size();
+  for(auto it = available_expressions.begin();
+      it != available_expressions.end();
+      it++)
+  {
+    if(!b.available_expressions.count(*it))
+    {
+      it = available_expressions.erase(it);
+      if(it == available_expressions.end())
+        break;
+    }
+  }
+
+  return size_before_intersection != available_expressions.size();
+}
+
+void cse_domaint::make_expression_available(const expr2tc &E)
+{
+  if(!E)
+    return;
+
+  // Logical expressions can be short-circuited
+  // So, only the first operand will be available
+  if(is_and2t(E) || is_or2t(E))
+  {
+    return;
+  }
+
+  if(is_if2t(E))
+  {
+    return;
+  }
+  // Skip sideeffects
+  if(is_sideeffect2t(E))
+    return;
+
+  // TODO: I hate floats
+  if(is_floatbv_type(E))
+    return;
+
+  // Did we check it already?
+  if(available_expressions.count(E))
+    return;
+
+  // Let's recursively make it available!
+  E->foreach_operand(
+    [this](const expr2tc &e) { make_expression_available(e); });
+
+  // TODO: LHS members should always be recomputed
+  if(is_with2t(E) || is_member2t(E) || is_dereference2t(E))
+    return;
+
+  available_expressions.insert(E);
+}
+
+bool cse_domaint::should_remove_expr(const expr2tc &taint, const expr2tc &E)
+  const
+{
+  if(!taint)
+    return false;
+
+  if(taint == E)
+    return true;
+
+  if(!E) // not sure when this can happen
+    return false;
+
+  bool result = false;
+  E->foreach_operand([this, &result, &taint](const expr2tc &e) {
+    result |= should_remove_expr(taint, e);
+  });
+  return result;
+}
+
+bool cse_domaint::should_remove_expr(const irep_idt &sym, const expr2tc &E)
+  const
+{
+  if(!E)
+    return false;
+
+  if(is_symbol2t(E) && to_symbol2t(E).thename == sym)
+    return true;
+
+  bool result = false;
+  E->foreach_operand([this, &result, &sym](const expr2tc &e) {
+    result |= should_remove_expr(sym, e);
+  });
+  return result;
+}
+
+void cse_domaint::havoc_symbol(const irep_idt &sym)
+{
+  std::vector<expr2tc> to_remove;
+  for(auto x : available_expressions)
+  {
+    if(should_remove_expr(sym, x))
+      to_remove.push_back(x);
+  }
+  for(auto x : to_remove)
+    available_expressions.erase(x);
+}
+
+void cse_domaint::havoc_expr(
+  const expr2tc &target,
+  const goto_programt::const_targett &i_it)
+{
+  if(is_dereference2t(target) && vsa != nullptr)
```

Comment:
Can it ever happen that `vsa == nullptr`?

Reply:
Yeah, if someone uses the avaliable expression domain without using the gcse class. I could convert it into an assertion... but I am not sure if that should be a hard requirement.